src/Raven.Server/Rachis/Leader.cs

```diff
@@ -710,19 +710,41 @@ private void EmptyQueue()
                         var djv = cmd.Command.ToJson(context);
                         var cmdJson = context.ReadObject(djv, "raft/command");
 
-                        var index = _engine.InsertToLeaderLog(context, Term, cmdJson, RachisEntryFlags.StateMachineCommand);
+                        if (_engine.HasHistoryLog(context, cmdJson, out var index, out var result, out var exception))
+                        {
+                            // if this command is already committed, we can skip it and notify the caller about it
+                            if (lastCommitted >= index) 
+                            {
+                                if (exception != null)
+                                {
+                                    cmd.Tcs.TrySetException(new Exception(exception));
```

Comment:
I don't like this, however. You are throwing `Exception`, and we probably should be throwing the actual exception type that was raised.

Reply:
Agree, but I keep the exception as a string and not sure how to convert it back to the original one.

Reply:
Keep the exception time and create it dynamically?